File: app/src/main/java/com/sortedqueue/programmercreek/adapter/UserProgramRecyclerAdapter.java
package com.sortedqueue.programmercreek.adapter;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.sortedqueue.programmercreek.CreekApplication;
import com.sortedqueue.programmercreek.R;
import com.sortedqueue.programmercreek.database.ProgramIndex;
import com.sortedqueue.programmercreek.database.UserProgramDetails;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Alok on 16/05/17.
 */

public class UserProgramRecyclerAdapter extends RecyclerView.Adapter<UserProgramRecyclerAdapter.ViewHolder> {


    private String accessSpecifier;
    private UserProgramClickListener adapterClickListner;
    private ArrayList<UserProgramDetails> userProgramDetailsArrayList;
    private Context context;

    private String userEmail;

    private Drawable likeDrawable;
    private Drawable unlikeDrawable;

    public interface UserProgramClickListener extends CustomProgramRecyclerViewAdapter.AdapterClickListner {
        void onLikeClicked( boolean isLiked, int position );
        void onShareClicked( int position );
    }

    public UserProgramRecyclerAdapter(Context context, String accessSpecifier, ArrayList<UserProgramDetails> modelArrayList, UserProgramClickListener adapterClickListner) {
        this.userProgramDetailsArrayList = modelArrayList;
        this.accessSpecifier = accessSpecifier;
        this.context = context;
        this.adapterClickListner = adapterClickListner;
        this.userEmail = CreekApplication.Companion.getCreekPreferences().getSignInAccount();
        this.likeDrawable = ContextCompat.getDrawable(context, R.drawable.ic_star_on);
        this.unlikeDrawable = ContextCompat.getDrawable(context, R.drawable.ic_star_off);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_user_program, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        UserProgramDetails userProgramDetails = userProgramDetailsArrayList.get(position);
        ProgramIndex programIndex = userProgramDetails.getProgramIndex();
        holder.titleTextView.setText(programIndex.getProgram_Language().toUpperCase() + " : " + programIndex.getProgram_Description());
        holder.subTitleTextView.setText(userProgramDetails.getPreview());
        holder.viewsTextView.setText(userProgramDetails.getViews() + " Views");
        holder.likesTextView.setText(userProgramDetails.getLikes() + " Likes");

        //holder.likesTextView.setCompoundDrawables( userProgramDetails.isLiked( userEmail ) ? likeDrawable : unlikeDrawable, null, null, null);
        holder.likesTextView.setCompoundDrawablesWithIntrinsicBounds( userProgramDetails.isLiked( userEmail ) ? likeDrawable : unlikeDrawable, null, null, null);
        holder.extrasLayout.setVisibility( accessSpecifier.equals("Favorites") ? View.GONE : View.VISIBLE );

    }

    @Override
    public int getItemCount() {
        return userProgramDetailsArrayList.size();
    }

    public UserProgramDetails getItemAtPosition(int position) {
        return userProgramDetailsArrayList.get(position);
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.titleTextView)
        TextView titleTextView;
        @BindView(R.id.subTitleTextView)
        TextView subTitleTextView;
        @BindView(R.id.viewsTextView)
        TextView viewsTextView;
        @BindView(R.id.likesTextView)
        TextView likesTextView;
        @BindView(R.id.shareImageView)
        ImageView shareImageView;
        @BindView(R.id.extrasLayout)
        LinearLayout extrasLayout;


        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            titleTextView.setOnClickListener(this);
            subTitleTextView.setOnClickListener(this);
            viewsTextView.setOnClickListener(this);
            likesTextView.setOnClickListener(this);
            shareImageView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            int position = getAdapterPosition();
            if (position != RecyclerView.NO_POSITION) {
                UserProgramDetails userProgramDetails = userProgramDetailsArrayList.get(position);
                if( v.getId() == R.id.likesTextView ) {
                    if( userProgramDetails.isLiked(userEmail) ) {
                        userProgramDetails.setLiked(false, userEmail);
                        adapterClickListner.onLikeClicked(false, position);
                    }
                    else {
                        userProgramDetails.setLiked(true, userEmail);
                        adapterClickListner.onLikeClicked(true, position);
                    }

                }
                else if( v.getId() == R.id.shareImageView ) {
                    adapterClickListner.onShareClicked(position);
                }
                else {
                    adapterClickListner.onItemClick(position);
                }
            }
        }
    }

}

File: app/src/main/java/com/sortedqueue/programmercreek/adapter/UserProgramRecyclerAdapter.kt
package com.sortedqueue.programmercreek.adapter

import android.content.Context
import android.graphics.drawable.Drawable
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import com.sortedqueue.programmercreek.CreekApplication
import com.sortedqueue.programmercreek.R
import com.sortedqueue.programmercreek.database.ProgramIndex
import com.sortedqueue.programmercreek.database.UserProgramDetails

import java.util.ArrayList




/**
 * Created by Alok on 16/05/17.
 */

class UserProgramRecyclerAdapter(private val context: Context, private val accessSpecifier: String, private val userProgramDetailsArrayList: ArrayList<UserProgramDetails>, private val adapterClickListner: UserProgramClickListener) : RecyclerView.Adapter<UserProgramRecyclerAdapter.ViewHolder>() {

    private val userEmail: String

    private val likeDrawable: Drawable
    private val unlikeDrawable: Drawable

    interface UserProgramClickListener : CustomProgramRecyclerViewAdapter.AdapterClickListner {
        fun onLikeClicked(isLiked: Boolean, position: Int)
        fun onShareClicked(position: Int)
    }

    init {
        this.userEmail = CreekApplication.creekPreferences!!.getSignInAccount()
        this.likeDrawable = ContextCompat.getDrawable(context, R.drawable.ic_star_on)
        this.unlikeDrawable = ContextCompat.getDrawable(context, R.drawable.ic_star_off)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.item_user_program, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val userProgramDetails = userProgramDetailsArrayList[position]
        val programIndex = userProgramDetails.programIndex
        holder.titleTextView!!.text = programIndex.program_Language.toUpperCase() + " : " + programIndex.program_Description
        holder.subTitleTextView!!.text = userProgramDetails.preview
        holder.viewsTextView!!.text = userProgramDetails.views.toString() + " Views"
        holder.likesTextView!!.text = userProgramDetails.likes.toString() + " Likes"

        //holder.likesTextView.setCompoundDrawables( userProgramDetails.isLiked( userEmail ) ? likeDrawable : unlikeDrawable, null, null, null);
        holder.likesTextView!!.setCompoundDrawablesWithIntrinsicBounds(if (userProgramDetails.isLiked(userEmail)) likeDrawable else unlikeDrawable, null, null, null)
        holder.extrasLayout!!.visibility = if (accessSpecifier == "Favorites") View.GONE else View.VISIBLE

    }

    override fun getItemCount(): Int {
        return userProgramDetailsArrayList.size
    }

    fun getItemAtPosition(position: Int): UserProgramDetails {
        return userProgramDetailsArrayList[position]
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        @BindView(R.id.titleTextView)
        internal var titleTextView: TextView? = null
        @BindView(R.id.subTitleTextView)
        internal var subTitleTextView: TextView? = null
        @BindView(R.id.viewsTextView)
        internal var viewsTextView: TextView? = null
        @BindView(R.id.likesTextView)
        internal var likesTextView: TextView? = null
        @BindView(R.id.shareImageView)
        internal var shareImageView: ImageView? = null
        @BindView(R.id.extrasLayout)
        internal var extrasLayout: LinearLayout? = null


        init {
            ButterKnife.bind(this, itemView)
            titleTextView!!.setOnClickListener(this)
            subTitleTextView!!.setOnClickListener(this)
            viewsTextView!!.setOnClickListener(this)
            likesTextView!!.setOnClickListener(this)
            shareImageView!!.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            val position = adapterPosition
            if (position != RecyclerView.NO_POSITION) {
                val userProgramDetails = userProgramDetailsArrayList[position]
                if (v.id == R.id.likesTextView) {
                    if (userProgramDetails.isLiked(userEmail)) {
                        userProgramDetails.setLiked(false, userEmail)
                        adapterClickListner.onLikeClicked(false, position)
                    } else {
                        userProgramDetails.setLiked(true, userEmail)
                        adapterClickListner.onLikeClicked(true, position)
                    }

                } else if (v.id == R.id.shareImageView) {
                    adapterClickListner.onShareClicked(position)
                } else {
                    adapterClickListner.onItemClick(position)
                }
            }
        }
    }

}